src/components/SwapWidget/SwapWidget.tsx

```diff
@@ -0,0 +1,247 @@
+import { useState, FormEvent } from 'react';
+import { toDecimalString } from "@airswap/utils";
+import { toAtomicString } from "@airswap/utils";
+import { BigNumber } from "ethers";
+import { findTokenByAddress } from "@airswap/metadata";
+import { parseUnits } from "ethers/lib/utils";
+import { useAppSelector, useAppDispatch } from "../../app/hooks";
+import {
+  SubmittedApproval,
+  selectTransactions,
+} from "../../features/transactions/transactionsSlice";
+import {
+  approve,
+  request,
+  take,
+  selectBestOrder,
+  selectOrdersStatus,
+} from "../../features/orders/ordersSlice";
+import { selectActiveTokens } from "../../features/metadata/metadataSlice";
+import { selectBalances } from "../../features/balances/balancesSlice";
+import { Web3Provider } from "@ethersproject/providers";
+import { useWeb3React } from "@web3-react/core";
+import { useMatomo } from "@datapunt/matomo-tracker-react";
+import { useTranslation } from "react-i18next";
+import Button from '../Button/Button';
+import TokenSelect from '../TokenSelect/TokenSelect';
+import Timer from "../../components/Timer/Timer";
+import Modal from "react-modal";
+import Card from '../Card/Card';
+import WalletProviderList from '../WalletProviderList/WalletProviderList';
+
+const floatRegExp = new RegExp("^[+-]?([0-9]+([.][0-9]*)?|[.][0-9]+)$");
+
+const SwapWidget = () => {
+  const [senderToken, setSenderToken] = useState<string>();
+  const [signerToken, setSignerToken] = useState<string>();
+  const [senderAmount, setSenderAmount] = useState("0.01");
+  const [isConnecting, setIsConnecting] = useState<boolean>(false);
+  const [showWalletList, setShowWalletList] = useState<boolean>(false);
+  const transactions = useAppSelector(selectTransactions);
+  const balances = useAppSelector(selectBalances);
+  const order = useAppSelector(selectBestOrder);
+  const ordersStatus = useAppSelector(selectOrdersStatus);
+  const dispatch = useAppDispatch();
+  const activeTokens = useAppSelector(selectActiveTokens);  
+  const { t } = useTranslation(["orders", "common", "wallet"]);
+  const { chainId, account, library, active } = useWeb3React<Web3Provider>();
+  const { trackEvent } = useMatomo();
+
+    
+
+  const getTokenDecimals = (tokenAddress: string) => {
+    for (const token of activeTokens) {
+      if (token.address === tokenAddress) return token.decimals;
+    }
+    return 18;
```

Comment:
This was a refactor of cody's code, I think I had signed off on the 18 decimal return because I assumed that was the standard, but you're right there can be a situation where we don't have access to token metadata. I'll make the change!

Reply:
How do you think it's best to approach a situation where `getTokenDecimals` returns `null`? ATM, it's used to parse `signerAmount` and determine `insufficientBalance`. On second thought, I don't know if we would ever run into a situation where we don't have a token's metadata? I do agree that assuming 18 decimals is wrong, but not exactly sure how we would handle a `null` case in the UI.